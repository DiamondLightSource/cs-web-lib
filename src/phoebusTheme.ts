import { createTheme } from "@mui/material/styles";

export const phoebusTheme = createTheme({
  palette: {
    primary: {
      main: "#D2D2D2",
      // light: currently calculated automatically by MUI
      // dark: currently calculated automatically by MUI
      contrastText: "#000000"
    }
  },
  typography: {
    fontFamily: "Liberation Sans",
    fontSize: 14,
    button: {
      textTransform: "none"
    }
  }
});
