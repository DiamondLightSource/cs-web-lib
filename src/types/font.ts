import { CSSProperties } from "react";

// assigned int values starting from zero
// also functions as a type
export enum FontStyle {
  Regular,
  Bold,
  Italic,
  BoldItalic
}

export class Font {
  // If no number is provided inherit from CSS.
  private size?: number;
  private style: FontStyle;
  private typeface: string;
  // TODO: name is currently unused
  private name?: string;

  public constructor(
    size?: number,
    style?: FontStyle,
    typeface?: string,
    name?: string
  ) {
    this.typeface = typeface ?? "Liberation sans";
    this.style = style ?? FontStyle.Regular;
    this.size = size;
    this.name = name;
  }

  /**
   * Returns a CSSProperties object containing all parameters defined by the
   * user and default properties for optional parameters not input by the user
   */
  public css(): CSSProperties {
    // Convert from px to REM by dividing pixels by default browser font size 16px
    // If a user has manually adjusted their browser font size, this might not be scaled
    // correctly, but given bob screens are pixel placed we shouldn't expect this to scale
    const fontSize = this.size ? `${this.size / 16}rem` : undefined;
    const fontWeight =
      this.style === FontStyle.Bold || this.style === FontStyle.BoldItalic
        ? "bold"
        : "normal";
    const fontStyle =
      this.style === FontStyle.Italic || this.style === FontStyle.BoldItalic
        ? "italic"
        : "normal";
    const style: CSSProperties = {
      // Fall back to sans-serif.
      fontFamily: `${this.typeface},sans-serif`,
      fontWeight: fontWeight,
      fontStyle: fontStyle
    };
    if (fontSize) {
      style.fontSize = fontSize;
    }
    return style;
  }
}
