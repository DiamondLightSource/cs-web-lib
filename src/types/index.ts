export { AbsolutePosition } from "./position";
export { RelativePosition } from "./position";
export { Color } from "./color";
export { PV } from "./pv";
export { Font, FontStyle } from "./font";
export { Border, BorderStyle } from "./border";
