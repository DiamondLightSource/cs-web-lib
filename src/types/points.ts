export class Point {
  public x: number;
  public y: number;

  public constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }
}

export class Points {
  public values: Array<Point>;

  public constructor(values: Array<Point>) {
    this.values = values;
  }
}
