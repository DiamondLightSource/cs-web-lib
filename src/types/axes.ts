import { Color } from "./color";
import { Font, FontStyle } from "./font";

export class Axis {
  public index: number;
  public autoScale?: boolean;
  public autoScaleThreshold?: number;
  public axisColor?: Color;
  public axisTitle?: string;
  public showGrid?: boolean;
  public gridColor?: Color;
  public dashGridLine?: boolean;
  public timeFormat?: number;
  public scaleFormat?: string;
  public scaleFont?: Font;
  public titleFont?: Font;
  public visible?: boolean;
  public logScale?: boolean;
  public leftBottomSide?: boolean;
  public maximum?: number;
  public minimum?: number;
  public yAxis?: boolean;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property. Uses same
   * default values as csstudio.opibuilder.xygraph.
   */
  public constructor(idx: number) {
    this.index = idx;
    this.autoScale = true;
    this.autoScaleThreshold = 0.95;
    this.axisColor = new Color("rgb(0, 0, 0");
    this.axisTitle = "";
    this.showGrid = false;
    this.gridColor = new Color("rgb(0, 0, 0");
    this.dashGridLine = true;
    this.timeFormat = 0;
    this.scaleFormat = "";
    this.scaleFont = new Font();
    this.titleFont = new Font(10, FontStyle.Bold);
    this.visible = true;
    this.logScale = false;
    this.leftBottomSide = true;
    this.minimum = 0;
    this.maximum = 100;
    // If axis has index 1 or above, assume y axis
    this.index > 0 ? (this.yAxis = true) : (this.yAxis = false);
  }
}

export class Axes {
  public count: number;
  public axisOptions: Axis[];

  public constructor(count: number, axes: Axis[]) {
    if (count !== axes.length) {
      throw new Error(
        `Count ${count} is not equal to number of axes ${axes.length}`
      );
    }
    this.count = count;
    this.axisOptions = axes;
  }
}
