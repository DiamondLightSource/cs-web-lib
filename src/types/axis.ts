import { Color } from "./color";
import { Font, FontStyle } from "./font";

export class Axis {
  public color: Color;
  public title: string;
  public showGrid: boolean;
  public visible: boolean;
  public logScale: boolean;
  public autoscale: boolean;
  public maximum: number;
  public minimum: number;
  public titleFont: Font;
  public scaleFont: Font;
  public onRight: boolean;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property. Uses same
   * default values as csstudio.opibuilder.xygraph.
   */
  public constructor({
    xAxis = false,
    color = new Color("rgb(0, 0, 0"),
    title = "",
    showGrid = false,
    visible = true,
    logScale = false,
    autoscale = false,
    minimum = 0,
    maximum = 100,
    titleFont = new Font(FontStyle.Bold),
    scaleFont = new Font(),
    onRight = false
  } = {}) {
    this.color = color;
    this.title = title || (xAxis ? "X" : "Y");
    this.showGrid = showGrid;
    this.visible = visible;
    this.logScale = logScale;
    this.autoscale = autoscale;
    this.minimum = minimum;
    this.maximum = maximum;
    this.titleFont = titleFont;
    this.scaleFont = scaleFont;
    this.onRight = onRight;
  }
}

export type Axes = Axis[];
