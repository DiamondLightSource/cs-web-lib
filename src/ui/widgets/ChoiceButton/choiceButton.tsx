import React, { useEffect, useState } from "react";

import { Widget } from "../widget";
import { PVComponent, PVWidgetPropType } from "../widgetProps";
import { registerWidget } from "../register";
import {
  BoolPropOpt,
  ColorPropOpt,
  FontPropOpt,
  InferWidgetProps,
  IntPropOpt,
  StringArrayPropOpt,
  StringPropOpt
} from "../propTypes";
import { DType } from "../../../types/dtypes";
// import classes from "./choiceButton.module.css";
import { writePv } from "../../hooks/useSubscription";
import {
  ToggleButton as MuiToggleButton,
  styled,
  ToggleButtonGroup
} from "@mui/material";
import { diamondTheme } from "../../../diamondTheme";
import { Color } from "../../../types";

const ChoiceButtonProps = {
  pvName: StringPropOpt,
  height: IntPropOpt,
  width: IntPropOpt,
  items: StringArrayPropOpt,
  selectedColor: ColorPropOpt,
  itemsFromPv: BoolPropOpt,
  foregroundColor: ColorPropOpt,
  backgroundColor: ColorPropOpt,
  horizontal: BoolPropOpt,
  enabled: BoolPropOpt,
  itemsfromPv: BoolPropOpt,
  font: FontPropOpt
};

export type ChoiceButtonComponentProps = InferWidgetProps<
  typeof ChoiceButtonProps
> &
  PVComponent;

const ToggleButton = styled(MuiToggleButton)({
  "&.MuiToggleButton-root": {
    textTransform: "none",
    overflow: "hidden",
    padding: 0,
    lineHeight: 1
  },
  "&.Mui-disabled": {
    cursor: "not-allowed",
    pointerEvents: "all !important"
  }
});

export const ChoiceButtonComponent = (
  props: ChoiceButtonComponentProps
): JSX.Element => {
  const {
    width = 100,
    height = 43,
    value = null,
    enabled = true,
    itemsFromPv = true,
    pvName,
    items = ["Item 1", "Item 2"],
    horizontal = true,
    foregroundColor = diamondTheme.palette.primary.contrastText,
    backgroundColor = diamondTheme.palette.primary.main,
    selectedColor = Color.fromRgba(200, 200, 200)
  } = props;
  const font = props.font?.css() ?? diamondTheme.typography;
  const [selected, setSelected] = useState(value?.getDoubleValue());

  // Use items from file, unless itemsFRomPv set
  let options = items;
  if (itemsFromPv && value?.display.choices) options = value?.display.choices;

  // This is necessary in order to set the initial label value
  // after connection to PV established, as setState cannot be
  // established inside a conditional, or called in the main body
  // of the component as it causes too many re-renders error
  useEffect(() => {
    if (value) {
      setSelected(value.getDoubleValue());
    }
  }, [value]);

  // Number of buttons to create
  const numButtons = options.length || 1;
  // Determine width and height of buttons if horizontal or vertically placed
  const buttonHeight = horizontal ? height : height / numButtons;
  const buttonWidth = horizontal ? width / numButtons : width;

  const handleChange = (
    event: React.MouseEvent<HTMLElement>,
    newSelect: number
  ) => {
    // Write to PV
    if (pvName) {
      writePv(pvName, new DType({ doubleValue: newSelect }));
    }
    setSelected(newSelect);
  };

  return (
    <ToggleButtonGroup
      exclusive
      fullWidth={true}
      disabled={!enabled}
      value={selected}
      onChange={handleChange}
      orientation={horizontal ? "horizontal" : "vertical"}
      sx={{
        height: height,
        width: width
      }}
    >
      {options
        .filter(item => typeof item === "string")
        .map((item, idx) => (
          <ToggleButton
            key={item}
            value={idx}
            sx={{
              minWidth: buttonWidth,
              minHeight: buttonHeight,
              fontFamily: font,
              color: foregroundColor.toString(),
              backgroundColor: backgroundColor.toString(),
              "&.Mui-selected, &.Mui-selected:hover, &:hover": {
                backgroundColor: selectedColor.toString()
              }
            }}
          >
            {item}
          </ToggleButton>
        ))}
    </ToggleButtonGroup>
  );
};

const ChoiceButtonWidgetProps = {
  ...ChoiceButtonProps,
  ...PVWidgetPropType
};

export const ChoiceButton = (
  props: InferWidgetProps<typeof ChoiceButtonWidgetProps>
): JSX.Element => <Widget baseWidget={ChoiceButtonComponent} {...props} />;

registerWidget(ChoiceButton, ChoiceButtonWidgetProps, "choicebutton");
