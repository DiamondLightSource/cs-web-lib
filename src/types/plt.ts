import { Axes, Axis } from "./axis";
import { Color } from "./color";
import { Font } from "./font";
import { Trace } from "./trace";

export class Plt {
  public title: string;
  public axes: Axes;
  public pvlist: Trace[];
  public backgroundColor: Color;
  public foregroundColor: Color;
  public scroll: boolean;
  public scrollStep: number;
  public updatePeriod: number;
  public start: string;
  public end: string;
  public showGrid: boolean;
  public titleFont: Font;
  public scaleFont: Font;
  public labelFont: Font;
  public legendFont: Font;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property.
   */
  public constructor({
    title = "",
    axes = [new Axis()],
    pvlist = [new Trace()],
    background = Color.WHITE,
    foreground = Color.BLACK,
    scroll = true,
    grid = false,
    scrollStep = 5,
    updatePeriod = 1,
    titleFont = new Font(),
    labelFont = new Font(),
    legendFont = new Font(),
    scaleFont = new Font(),
    start = "1 minute",
    end = "now"
  } = {}) {
    this.backgroundColor = background;
    this.foregroundColor = foreground;
    this.title = title;
    this.scroll = scroll;
    this.scrollStep = scrollStep;
    this.axes = axes;
    this.pvlist = pvlist;
    this.updatePeriod = updatePeriod;
    this.showGrid = grid;
    this.start = start;
    this.end = end;
    this.titleFont = titleFont;
    this.scaleFont = scaleFont;
    this.labelFont = labelFont;
    this.legendFont = legendFont;
  }
}
