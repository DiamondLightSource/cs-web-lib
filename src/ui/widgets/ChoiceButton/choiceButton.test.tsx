import React from "react";
import { ChoiceButtonComponent } from "./choiceButton";
import { fireEvent, render } from "@testing-library/react";
import { DDisplay, DType } from "../../../types/dtypes";
import { Color, Font } from "../../../types";
import { ThemeProvider } from "@mui/material";
import { phoebusTheme } from "../../../phoebusTheme";

const ChoiceButtonRenderer = (choiceButtonProps: any): JSX.Element => {
  return (
    <ThemeProvider theme={phoebusTheme}>
      <ChoiceButtonComponent {...choiceButtonProps} />
    </ThemeProvider>
  );
};

describe("<ChoiceButton />", (): void => {
  test("it renders ChoiceButton with default props", (): void => {
    const choiceButtonProps = {
      value: null
    };
    const { getAllByRole } = render(ChoiceButtonRenderer(choiceButtonProps));
    const buttons = getAllByRole("button") as Array<HTMLButtonElement>;

    expect(buttons[0].textContent).toEqual("Item 1");
    expect(buttons[1].textContent).toEqual("Item 2");

    buttons.forEach(button => {
      expect(button).toHaveStyle({
        "background-color": "rgb(210, 210, 210)",
        height: "43px",
        width: "50px",
        "font-size": "14px"
      });
    });
  });

  test("pass props to widget", (): void => {
    const choiceButtonProps = {
      value: new DType({ doubleValue: 0 }),
      width: 60,
      height: 140,
      font: new Font(12),
      items: ["Choice", "Option", "Setting", "Custom"],
      horizontal: false,
      backgroundColor: Color.fromRgba(20, 20, 200),
      selectedColor: Color.fromRgba(10, 60, 40),
      itemsFromPv: false,
      enabled: false
    };
    const { getAllByRole } = render(ChoiceButtonRenderer(choiceButtonProps));
    const buttons = getAllByRole("button") as Array<HTMLButtonElement>;

    expect(buttons.length).toEqual(4);
    expect(buttons[2].textContent).toEqual("Setting");
    expect(buttons[3]).toHaveProperty("disabled", true);

    expect(buttons[0]).toHaveStyle({
      "background-color": "rgb(20, 20, 200)",
      "font-size": "0.75rem"
    });

    expect(buttons[3]).toHaveStyle({
      cursor: "not-allowed",
      height: "35px"
    });
  });

  test("pass props to widget, using itemsFromPv", (): void => {
    const choiceButtonProps = {
      value: new DType(
        { doubleValue: 0 },
        undefined,
        undefined,
        new DDisplay({ choices: ["hi", "Hello"] })
      ),
      items: ["one", "two", "three"],
      horizontal: false,
      itemsFromPv: true,
      enabled: true
    };
    const { getAllByRole } = render(ChoiceButtonRenderer(choiceButtonProps));
    const buttons = getAllByRole("button") as Array<HTMLButtonElement>;

    expect(buttons.length).toEqual(2);
    // First button is selected therefore different color and box shadow
    expect(buttons[0].textContent).toEqual("hi");
    expect(buttons[1].textContent).toEqual("Hello");
  });

  test("selecting a button", (): void => {
    const choiceButtonProps = {
      value: null
    };
    const { getAllByRole } = render(ChoiceButtonRenderer(choiceButtonProps));
    const buttons = getAllByRole("button") as Array<HTMLButtonElement>;

    expect(buttons[0].textContent).toEqual("Item 1");
    expect(buttons[1].textContent).toEqual("Item 2");

    expect(buttons[0]).toHaveStyle({
      "background-color": "rgb(210, 210, 210)"
    });
    fireEvent.click(buttons[0]);
    expect(buttons[0]).toHaveStyle({
      "background-color": "rgb(210, 210, 210)"
    });
  });
});
