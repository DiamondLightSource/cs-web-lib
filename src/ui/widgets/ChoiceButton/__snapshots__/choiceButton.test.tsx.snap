// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<BoolButton /> > it renders ChoiceButton with default props 1`] = `
<DocumentFragment>
  <div
    class="MuiToggleButtonGroup-root MuiToggleButtonGroup-horizontal MuiToggleButtonGroup-fullWidth css-4ybkxl-MuiToggleButtonGroup-root"
    role="group"
  >
    <button
      aria-pressed="false"
      class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-firstButton css-1269rce-MuiButtonBase-root-MuiToggleButton-root"
      tabindex="0"
      type="button"
      value="0"
    >
      Item 1
    </button>
    <button
      aria-pressed="false"
      class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-lastButton css-1269rce-MuiButtonBase-root-MuiToggleButton-root"
      tabindex="0"
      type="button"
      value="1"
    >
      Item 2
    </button>
  </div>
</DocumentFragment>
`;

exports[`<BoolButton /> > pass props to widget 1`] = `
<DocumentFragment>
  <div
    class="MuiToggleButtonGroup-root MuiToggleButtonGroup-vertical MuiToggleButtonGroup-fullWidth css-pqsrz6-MuiToggleButtonGroup-root"
    role="group"
  >
    <button
      aria-pressed="true"
      class="MuiButtonBase-root Mui-disabled MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedVertical Mui-disabled MuiToggleButton-root Mui-selected Mui-disabled MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-firstButton css-tzwn7e-MuiButtonBase-root-MuiToggleButton-root"
      disabled=""
      tabindex="-1"
      type="button"
      value="0"
    >
      Choice
    </button>
    <button
      aria-pressed="false"
      class="MuiButtonBase-root Mui-disabled MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedVertical Mui-disabled MuiToggleButton-root Mui-disabled MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-middleButton css-tzwn7e-MuiButtonBase-root-MuiToggleButton-root"
      disabled=""
      tabindex="-1"
      type="button"
      value="1"
    >
      Option
    </button>
    <button
      aria-pressed="false"
      class="MuiButtonBase-root Mui-disabled MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedVertical Mui-disabled MuiToggleButton-root Mui-disabled MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-middleButton css-tzwn7e-MuiButtonBase-root-MuiToggleButton-root"
      disabled=""
      tabindex="-1"
      type="button"
      value="2"
    >
      Setting
    </button>
    <button
      aria-pressed="false"
      class="MuiButtonBase-root Mui-disabled MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedVertical Mui-disabled MuiToggleButton-root Mui-disabled MuiToggleButton-fullWidth MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-lastButton css-tzwn7e-MuiButtonBase-root-MuiToggleButton-root"
      disabled=""
      tabindex="-1"
      type="button"
      value="3"
    >
      Custom
    </button>
  </div>
</DocumentFragment>
`;
